core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -283,6 +300,48 @@ impl LatestUnprocessedVotes {
             .and_then(|l| l.read().unwrap().timestamp())
     }
 
+    #[cfg(test)]
+    pub(crate) fn set_staked_nodes(&self, staked_nodes: &[Pubkey]) {
+        let staked_nodes: HashMap<Pubkey, u64> =
+            staked_nodes.iter().map(|pk| (*pk, 1u64)).collect();
+        *self.cached_staked_nodes.write().unwrap() = Arc::new(staked_nodes);
+    }
+
+    fn weighted_random_order_by_stake(&self) -> impl Iterator<Item = Pubkey> {
+        // Efraimidis and Spirakis algo for weighted random sample without replacement
+        let staked_nodes = self.cached_staked_nodes.read().unwrap();
+        let latest_votes_per_pubkey = self.latest_votes_per_pubkey.read().unwrap();
+        let mut pubkey_with_weight: Vec<(f64, Pubkey)> = latest_votes_per_pubkey
+            .keys()
+            .filter_map(|&pubkey| {
+                let stake = staked_nodes.get(&pubkey).copied().unwrap_or(0);
+                if stake == 0 {
+                    None // Ignore votes from unstaked validators
+                } else {
+                    Some((thread_rng().gen::<f64>().powf(1.0 / (stake as f64)), pubkey))
+                }
+            })
+            .collect::<Vec<_>>();
+        pubkey_with_weight.sort_by(|(w1, _), (w2, _)| w1.partial_cmp(w2).unwrap());
+        pubkey_with_weight.into_iter().map(|(_, pubkey)| pubkey)
+    }
+
+    /// Recache the staked nodes based on a bank from the new epoch.
+    /// This should only be run by the TPU vote thread
+    pub(super) fn cache_epoch_boundary_info(&self, bank: &Bank) {
+        if bank.epoch() <= self.current_epoch.load(Ordering::Relaxed) {
```

Comment:
When would we expect bank epoch to be less than what we cached here?

Reply:
I think it's possible in an edge case duplicate scenario, since the bank here is the latest `working_bank` from `bank_forks`.

You could have `new_epoch_bank` be processed and `current_epoch` updated. Then `new_epoch_bank` is dumped due to it being a duplicate, meanwhile we receive the last bank from the previous epoch. In that case I believe the new `working_bank` would be from the previous epoch and thus `bank.epoch()` would now be less than `self.current_epoch`

Reply:
This is just leader side inclusion so it shouldn't be able to diverge consensus. In the worst case a newly staked voter might have their first vote thrown out because the cached epoch stakes was not updated in time, but that would be quickly resolved once we process blocks in Epoch 2.

The reverse (a newly unstaked voter) will not cause us problems because 0 stake votes are not included when packing the block.